File: 290-word-pattern/word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if(pattern.length() != words.length){
            return false;
        }

        Map<Character, String> map = new HashMap<>();
        Map<String, Character> reversemap = new HashMap<>();

        for(int i=0; i<pattern.length(); i++){
            char ch = pattern.charAt(i);
            String word = words[i];

            if(map.containsKey(ch)){
                if(!map.get(ch).equals(word)){
                    return false;
                }
            }
            else{
                map.put(ch,word);
            }

            if(reversemap.containsKey(word)){
                if(reversemap.get(word) != ch){
                    return false;
                }
            }else{
                reversemap.put(word,ch);
            }
        }
        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()

        if len(pattern) != len(words):
            return False

        pattern_map = {}
        word_map = {}

        for i in range(len(pattern)):
            ch = pattern[i]
            word = words[i]

            if ch in pattern_map:
                if pattern_map[ch] != word:
                    return False
            else:
                pattern_map[ch] = word

            if word in word_map:
                if word_map[word] != ch:
                    return False
            else:
                word_map[word] = ch

        return True
        
         